machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,131 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuild describes a build process managed by the MCO
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuild struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machine os build
+	// +kubebuilder:validation:Required
+	Spec MachineOSBuildSpec `json:"spec"`
+
+	// status describes the lst observed state of this machine os build
+	// +optional
+	Status MachineOSBuildStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuildList describes all of the Builds on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineOSObjectReference `json:"machineConfigPool"`
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSObjectReference `json:"renderedMachineOSImage"`
+	// buildInputs is where user options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig string `json:"currentConfig"`
```

Comment:
Good question, I don't have as much background but we should make sure there's a clear source of truth for when a build needs to be triggered, and this delta between current and desired sounds like it needs to be calculated by a controller, not a user, so wouldn't expect these in spec.

Reply:
desired will be necessary. We need something to trigger the build. Similarly to how currently we query the MCP and look at desired/current to trigger a build, something on the MachineOSBuild will need to do that now. Otherwise we will need sync loops in the MachineOSBuilder for MCPs and MOSB, I'd like to remove the MCP sync loop if possible.

Reply:
What and who would set the desired? Is the desired likely to change frequently?

Reply:
question: Shouldn't we make this a `RenderedMachineConfigReference`? Also, do we still need the `CurrentConfig` / `DesiredConfig` distinction here? Maybe it should just be `RenderedMachineConfig`?

Reply:
I am implementing this in MCO, we do need currentConfig for a few parts in the buildController. I can go dig them up if need be